### PR TITLE
Allow factory state to be set

### PR DESCRIPTION
**Pull request by kielabokkie**
This PR allows you to pass factory states via the config like below:

```php
'mailables' => [
    UserVerified::class => [
        ['class' => User::class, 'states' => ['verified']]
    ]
]
```

### CODE REVIEW

**Comment by kielabokkie** [comment]
I have a few Mailables that need my model to be in a specific state. It makes the mailables configuration array a bit less clean but more flexible. Let me know what you think @introwit

File: database/factories/TestFactory.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+use JoggApp\MailViewer\Tests\Stubs\Models\Test;
+
+/** @var \Illuminate\Database\Eloquent\Factory $factory */
+
+$factory->define(Test::class, function () {
```
Comment (introwit) [inline database/factories/TestFactory.php]: I think we can move this file to the `tests` directory 😄

File: src/MailViewer.php [hunk 2]
```diff
@@ -42,9 +42,20 @@ public static function find(string $mail)
                 $args = [];
 
                 foreach ($dependencies as $dependency) {
+                    $factoryStates = [];
+
+                    if (is_array($dependency)) {
+                        // Check if the dependency array contains factory states
```
Comment (introwit) [inline src/MailViewer.php]: I feel we can remove this comment since the code is readable enough 😄

File: src/MailViewer.php [hunk 3]
```diff
@@ -42,9 +42,20 @@ public static function find(string $mail)
...
+                        if (in_array('states', array_keys($dependency)) === true) {
```
Comment (introwit) [inline src/MailViewer.php]: We can remove that `=== true` since `in_array` returns a boolean iirc 🙂

File: src/MailViewer.php [hunk 4]
```diff
@@ -42,9 +42,20 @@ public static function find(string $mail)
...
+                            // Set the states and actual dependency class
```
Comment (introwit) [inline src/MailViewer.php]: I guess we can remove this comment too 😄

File: src/MailViewer.php [hunk 5]
```diff
@@ -42,9 +42,20 @@ public static function find(string $mail)
...
+                            $factoryStates = $dependency['states'];
+                            $dependency = $dependency['class'];
+                        }
+                    }
+
                     if (is_string($dependency) && class_exists($dependency)) {
                         if (isset($eloquentFactory[$dependency])) {
-                            $args[] = factory($dependency)->create();
+                            $args[] = factory($dependency)->states($factoryStates)->make();
```
Comment (introwit) [inline src/MailViewer.php:56]: Curious to know why `make()` over `create()` 😄

File: src/MailViewer.php [hunk 6]
```diff
@@ -70,6 +81,14 @@ public static function prepareMails(array $mailables): array
             $givenParameters = [];
 
             foreach ($dependencies as $dependency) {
+                if (is_array($dependency)) {
+                    // Check if the dependency array contains factory states
+                    if (in_array('states', array_keys($dependency)) === true) {
```
Comment (introwit) [inline src/MailViewer.php]: Same as above for the comment and `=== true` 🙂

File: src/MailViewer.php [hunk 7]
```diff
@@ -83,8 +102,7 @@ public static function prepareMails(array $mailables): array
 
             if ($constructorParameters !== $givenParameters) {
                 throw new Exception(
-                    "The arguments passed for {$mailable} in the config/mailviewer.php file do not match with the constructor 
-                    params of the {$mailable} class or the constructor params of the {$mailable} class aren't typehinted"
+                    "The arguments passed for {$mailable} in the config/mailviewer.php file do not match with the constructor params of the {$mailable} class or the constructor params of the {$mailable} class aren't typehinted"
```
Comment (introwit) [inline src/MailViewer.php]: Can you revert this change so that we adhere to PSR2's guideline for length of the line being 120 chars max. Also, we won't need to horizontally scroll when interacting with this file on GitHub if done 😄

File: src/MailViewerServiceProvider.php [hunk 8]
```diff
@@ -15,6 +16,15 @@ public function boot()
         $this->loadRoutesFrom(__DIR__ . '/../routes/web.php');
 
         $this->loadViewsFrom(__DIR__ . '/../views', 'mailviewer');
+
+        if (app()->environment() === 'testing') {
```
Comment (introwit) [inline src/MailViewerServiceProvider.php:20]: Curious to know what is the benefit of this container binding since it might effect user's own testing env

**Review by introwit — CHANGES_REQUESTED** [review]

**Comment by introwit** [comment]
@kielabokkie Thank you so much for the contribution Wouter ❤️ 🙂 Just some minor stuff and few questions out of curiosity, rest looks great 👌 Also, a small request can you update the comment in the config file to document the new behaviour & feature if possible? Thank you again 😄

File: src/MailViewerServiceProvider.php (hunk 8 above)
Comment (kielabokkie) [reply src/MailViewerServiceProvider.php:20]: I had to do this to make sure the `TestFactory.php` class would be loaded when running the unit tests. Laravel sets the default environment to `testing` when running unit tests so shouldn't cause a problem for anyone's testing env. See [the code on Laravel's repo](https://github.com/laravel/framework/blob/763b64a43ebb6042e463aab4214d4cc9722147be/src/Illuminate/Foundation/Application.php#L530).

File: src/MailViewer.php (hunk 5 above)
Comment (kielabokkie) [reply src/MailViewer.php:56]: The first reason was the tests, the create function will try to save the Test model in the database as it's a Illuminate model. I can make it some general model that doesn't save to the database to make the test pass too. The other reason is that you probably don't want to save the records to the database when you are just viewing the email via the `/mails` page, especially on a production environment ;) I actually didn't realise this was happening until I was editing this code.

**Comment by kielabokkie** [comment]
@introwit I've updated the PR based on your feedback. Let me know if there is anything else I can do.

**Review by introwit — APPROVED** [review]

**Comment by introwit** [comment]
Thank you so much @kielabokkie 🙂 Is it okay if I tag you in future conversations related to this feature, if any comes up? I will merge this and tag a new minor release `v2.0.3` 😄 Great work Wouter ❤️

**Comment by kielabokkie** [comment]
@introwit Yes of course, that's not a problem at all. Thank you for the awesome package 👍

**Comment by introwit** [comment]
@kielabokkie thank you for the contribution Wouter 🙂 Have tagged the new release [v2.1.0](https://github.com/JoggApp/laravel-mail-viewer/releases/tag/v2.1.0) and added credits to you in the [changelog](https://github.com/JoggApp/laravel-mail-viewer/blob/master/CHANGELOG.md) & [release](https://github.com/JoggApp/laravel-mail-viewer/releases/tag/v2.1.0). Great work 👌

**Comment by kielabokkie** [comment]
@introwit I just realised an issue with the environment check in the service provider. All the unit tests that use factories on my project are now failing because it's only loading the factory that is part of this package. I'll have a quick look to see how I can fix this.

**Comment by introwit** [comment]
@kielabokkie if it takes too long just send a PR that removes that test factory & that binding, since it's just needed for the test anyways 🙂
